### PR TITLE
Mint mutation

### PR DESCRIPTION
**Pull request by 0xjocke**
This is based on #382. So maybe have a look at that one first.

This PR have minting working. 
- I changes the `Mint` component to the container pattern we use elsewhere. So transaction logic got moved from the page to the container component.
- useMintMutation got refactored to work with the new gasHook. I reuse `populateTransaction`  function  and then send the transaction with `signer.sendTransaction(populatedTransaction)`
- Use unstaked snx rather thatn snx balance in the mint ui
-  Added transaction modal
Screen Shot 2022-10-05 at 11 21 38 am
Screen Shot 2022-10-05 at 11 21 30 am

### CODE REVIEW

**Comment by bachstatter** [comment]
I now there's no tests. Testing this requires lots of mocking so I think I would like to maybe build out burn and see if architecture changes before adding tests

**Comment by bachstatter** [comment]
Next steps. 
1. Make sure errors gets handled/ displayed
2. Etherscan url
3. Fix up burn

**Review by jmzwar — APPROVED** [review]

File: v2/components/Mint/Mint.tsx [hunk 1]
```diff
@@ -1,221 +1,283 @@
...
+          fontWeight="black"
+          mt={4}
+          w="100%"
+          onClick={() => onSubmit()}
+          disabled={mintAmountSNX === ''}
+        >
+          Mint
+        </Button>
+      </Box>
+      <TransactionModal
```
Comment (jmzwar) [inline v2/components/Mint/Mint.tsx:194]: No icon here? - we can always add later

File: v2/components/Mint/Mint.tsx [hunk 2]
```diff
@@ -1,221 +1,283 @@
...
+        isLoading={isLoading}
+        exchangeRate={exchangeRate}
+        mintAmountSNX={mintAmountSNX}
+        onMintAmountSNXChange={setMintAmountSNX}
+        unstakedSnx={unstakedSnx.toNumber()}
+        susdBalance={synthsData?.balancesMap.sUSD?.balance.toNumber()}
+        onSubmit={() => {
+          mutate(undefined, {
+            onSuccess: () => {
+              refetchSynthBalances();
```
Comment (jmzwar) [inline v2/components/Mint/Mint.tsx:272]: I believe there is a better way to do this -  
```
await queryClient.refetchQueries(['synths'], { active: true })
```
Or whatever the key for synths is

Comment (bachstatter) [reply v2/components/Mint/Mint.tsx:272]: Hmm I could, but they key is a bit dynamic
`['v2walletBalances', 'synths', networkId, walletAddress]`
So I would need to grab `networkId`, `walletAddress` and `queryClient` from context. So it seems easier just to call refetch?

File: v2/components/Mint/Mint.tsx (hunk 1 above)
Comment (bachstatter) [reply v2/components/Mint/Mint.tsx:194]: Missed that, will add

File: v2/components/Mint/Mint.tsx (hunk 2 above)
Comment (jmzwar) [reply v2/components/Mint/Mint.tsx:272]: I'm not sure I get what you're asking - we just want to update all synth balances after the call right - so any mounted hook that has the synths key should refetch?

I.e balances et al

Comment (jmzwar) [reply v2/components/Mint/Mint.tsx:272]: Does it have to be all the keys?

Comment (bachstatter) [reply v2/components/Mint/Mint.tsx:272]: ah maybe that how it works, I thought I needed all keys..

File: v2/components/EthGasPriceEstimator/EthGasPriceEstimator.tsx [hunk 3]
```diff
@@ -21,7 +21,7 @@ export const EthGasPriceEstimatorUi: React.FC<{
         <Menu closeOnSelect={true}>
           {() => (
             <>
-              <MenuButton>
+              <MenuButton data-testid="transactionPrice">
```
[the remark this answers is not in the document]
Comment (bachstatter) [reply v2/components/EthGasPriceEstimator/EthGasPriceEstimator.tsx:24]: yep sounds good to me

File: v2/lib/useMintMutation/useMintMutation.ts [hunk 4]
```diff
@@ -1,80 +1,86 @@
...
+      return Synthetix.populateTransaction[method](mintArgs.amount, {
+        gasLimit: Synthetix.estimateGas[method](mintArgs.amount),
+      });
+    }
+    if (method === 'issueMaxSynths') {
+      return Synthetix.populateTransaction[method]({
+        gasLimit: Synthetix.estimateGas[method](),
+      });
+    }
+    if (delegateAddress && method === 'issueSynthsOnBehalf') {
```
[the remark this answers is not in the document]
Comment (bachstatter) [reply v2/lib/useMintMutation/useMintMutation.ts:39]: Yea I agree. I started out with a more "dynamic" way but realised that all calls are in fact different. So now the `method` logic is unnecessary. Will update

File: v2/lib/useMintMutation/useMintMutation.ts [hunk 5]
```diff
@@ -84,10 +90,10 @@ export function useMintMutation(delegateAddress?: DelegationWallet | null) {
         setTimeout(() => dispatch({ type: 'settled' }), 1000);
```
[the remark this answers is not in the document]
Comment (bachstatter) [reply v2/lib/useMintMutation/useMintMutation.ts:90]: yepyep.. Might tackle this in the next PR where I handle errors

File: v2/components/EthGasPriceEstimator/EthGasPriceEstimator.cy.tsx [hunk 6]
```diff
@@ -21,14 +20,12 @@ describe('CRatioHealthCard.cy.tsx', () => {
...
-          ethPriceRate={wei(1000)}
-          gasPrice={{ gasPrice: BigNumber.from(1000000) }}
-          gasLimit={BigNumber.from(2000000)}
-          optimismLayerOneFees={wei(0.000001)}
+          transactionFee={wei(0.003)}
         />
       </Box>
     );
-    cy.contains('p', '$0.003').should('be.visible');
+    cy.get('[data-testid=transactionPrice]').should('have.text', '$0.003');
```
[the remark this answers is not in the document]
Comment (bachstatter) [reply v2/components/EthGasPriceEstimator/EthGasPriceEstimator.cy.tsx]: Yep implemented

File: v2/components/Mint/Mint.tsx (hunk 2 above)
Comment (bachstatter) [reply v2/components/Mint/Mint.tsx]: Fixed

File: v2/components/EthGasPriceEstimator/EthGasPriceEstimator.cy.tsx [hunk 7]
```diff
@@ -1,9 +1,8 @@
 import { Box } from '@chakra-ui/react';
 import { EthGasPriceEstimatorUi } from './EthGasPriceEstimator';
 import { wei } from '@synthetixio/wei';
-import { BigNumber } from '@ethersproject/bignumber';
 
-describe('CRatioHealthCard.cy.tsx', () => {
+describe('EthGasPriceEstimator.cy.tsx', () => {
```
[the remark this answers is not in the document]
Comment (bachstatter) [reply v2/components/EthGasPriceEstimator/EthGasPriceEstimator.cy.tsx:5]: haha haven't seen that before. Lets do it

File: v2/ui/content/V2Mint.tsx [hunk 8]
```diff
@@ -250,14 +235,7 @@ const V2Mint = () => {
...
         )}
-        <Mint
-          snxBalance={debtData?.collateral || wei(0)}
-          susdBalance={synthsData?.balancesMap['sUSD']?.balance || wei(0)}
-          gasPrice={wei(20)}
-          exchangeRate={exchangeRate}
-          isLoading={isLoading}
-          onSubmit={onSubmit}
-        />
+        <Mint delegateWalletAddress={delegateWallet?.address} />
```
[the remark this answers is not in the document]
Comment (bachstatter) [reply v2/ui/content/V2Mint.tsx:238]: Haven't moved over the delegate state from recoil, so keeping it like this until we have a delegateWallet context

**Review by noisekit — APPROVED** [review]

**Pull request merged**